### PR TITLE
[MemProf] Refactor and clean up edge removal

### PR DESCRIPTION
**Pull request by teresajohnson**
Add helper for removing an edge from the graph, and for checking if an
edge has been removed from the graph, and then update code to use those
consistently for removal and during edge iteration, respectively. Also
fix a couple of places that were incorrectly iterating over edge lists
that could in theory be updated during the iteration.

### CODE REVIEW

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp [hunk 1]
```diff
@@ -417,6 +417,29 @@ class CallsiteContextGraph {
 
     DenseSet<uint32_t> &getContextIds() { return ContextIds; }
 
+    // Helper to clear the fields of this edge when we are removing it from the
+    // graph.
+    void clear() {
```
Comment (snehasish) [inline llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:422]: Mark these small functions as inline?

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp [hunk 2]
```diff
@@ -920,15 +952,42 @@ void CallsiteContextGraph<DerivedCCG, FuncTy, CallTy>::ContextNode::
...
+    ContextEdge *Edge, EdgeIter *EI, bool CalleeIter) {
+  assert(!EI || (*EI)->get() == Edge);
+  // Save the Caller and Callee pointers so we can erase Edge from their edge
+  // lists after clearing Edge below. We do the clearing first in case it is
+  // destructed after removing from the edge lists (if those were the last
+  // shared_ptr references to Edge).
+  auto *Callee = Edge->Callee;
+  auto *Caller = Edge->Caller;
+
+  // Make sure the edge fields are emptied/nulled out so we can properly detect
```
Comment (snehasish) [inline llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:966]: nit: s/emptied\/nulled/cleared/

**Review by snehasish — APPROVED** [review]
lgtm

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp (hunk 2 above)
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp]: done

**Pull request merged**